src/combinatorics/simple_graph/subgraph.lean

```diff
@@ -0,0 +1,313 @@
+def is_subgraph (x y : subgraph G) : Prop := x.verts ⊆ y.verts ∧ ∀ {v w : V}, x.adj v w → y.adj v w
+
+instance : bounded_lattice (subgraph G) :=
+{ le := is_subgraph,
+  sup := union,
+  inf := inter,
+  top := top,
+  bot := bot,
+  le_refl := λ x, ⟨rfl.subset, λ _ _ h, h⟩,
+  le_trans := λ x y z hxy hyz, ⟨hxy.1.trans hyz.1, λ _ _ h, hyz.2 (hxy.2 h)⟩,
+  le_antisymm := begin
+    intros x y hxy hyx,
+    ext1 v,
+    exact set.subset.antisymm hxy.1 hyx.1,
+    ext v w,
+    exact iff.intro (λ h, hxy.2 h) (λ h, hyx.2 h),
+  end,
+  le_top := λ x, ⟨set.subset_univ _, (λ v w h, x.adj_sub h)⟩,
+  bot_le := λ x, ⟨set.empty_subset _, (λ v w h, false.rec _ h)⟩,
+  sup_le := λ x y z hxy hyz,
+            ⟨set.union_subset hxy.1 hyz.1,
+              (λ v w h, h.cases_on (λ h, hxy.2 h) (λ h, hyz.2 h))⟩,
+  le_sup_left := λ x y, ⟨set.subset_union_left x.verts y.verts, (λ v w h, or.inl h)⟩,
+  le_sup_right := λ x y, ⟨set.subset_union_right x.verts y.verts, (λ v w h, or.inr h)⟩,
+  le_inf := λ x y z hxy hyz, ⟨set.subset_inter hxy.1 hyz.1, (λ v w h, ⟨hxy.2 h, hyz.2 h⟩)⟩,
+  inf_le_left := λ x y, ⟨set.inter_subset_left x.verts y.verts, (λ v w h, h.1)⟩,
+  inf_le_right := λ x y, ⟨set.inter_subset_right x.verts y.verts, (λ v w h, h.2)⟩ }
+
+/-- The top of the `subgraph G` lattice is equivalent to the graph itself. -/
+def top_equiv : (⊤ : subgraph G).coe ≃g G :=
+{ to_fun := λ v, ↑v,
+  inv_fun := λ v, ⟨v, by tidy⟩,
+  left_inv := by tidy,
+  right_inv := by tidy,
+  map_rel_iff' := by tidy }
+
+/-- The bottom of the `subgraph G` lattice is equivalent to the empty graph on the empty
+vertex type. -/
+def bot_equiv : (⊥ : subgraph G).coe ≃g empty_graph empty :=
+{ to_fun := λ v, false.elim v.property,
+  inv_fun := λ v, begin cases v, end,
+  left_inv := by tidy,
+  right_inv := by tidy,
+  map_rel_iff' := by tidy }
+
+/-- Given two subgraphs, one a subgraph of the other, there is an induced injective homomorphism of
+the subgraphs as graphs. -/
+def map {x y : subgraph G} (h : x ≤ y) : x.coe →g y.coe :=
+{ to_fun := λ v, ⟨↑v, and.left h v.property⟩,
+  map_rel' := λ v w hvw, h.2 hvw }
+
+lemma map.injective {x y : subgraph G} (h : x ≤ y) : function.injective (map h) :=
+λ v w h, by { simp only [map, rel_hom.coe_fn_mk, subtype.mk_eq_mk] at h, exact subtype.ext h }
+
+/-- There is an induced injective homomorphism of a subgraph of `G` into `G`. -/
+def map_top (x : subgraph G) : x.coe →g G :=
+{ to_fun := λ v, v,
+  map_rel' := λ v w hvw, x.adj_sub hvw }
+
+lemma map_top.injective {x : subgraph G} : function.injective x.map_top :=
+λ v w h, subtype.ext h
+
+@[simp]
+lemma map_top_to_fun {x : subgraph G} (v : x.verts) : x.map_top v = v := rfl
+
+lemma neighbor_set_subset_of_subgraph {x y : subgraph G} (h : x ≤ y) (v : V) :
+  x.neighbor_set v ⊆ y.neighbor_set v :=
+λ w h', h.2 h'
+
+instance neighbor_set.decidable_pred (G' : subgraph G) [h : decidable_rel G'.adj] (v : V) :
+  decidable_pred (G'.neighbor_set v) := h v
```

Comment:
This needs a ` ∈`

Reply:
```suggestion
  decidable_pred (∈ G'.neighbor_set v) := h v
```

Reply:
Oh, that's good this was fixed -- I was meaning to get around to that.  Thanks.

I pushed the changes.